gnus-recent.el

```diff
@@ -165,60 +216,282 @@ Warn if RECENT can't be deconstructed as expected."
   "Update RECENT Gnus article in `gnus-recent--articles-list'.
 The Gnus article has moved to group TO-GROUP."
-  (cl-nsubstitute (list (cl-first recent) (cl-second recent) to-group)
-                  recent
-                  gnus-recent--articles-list
-                  :test 'equal :count 1))
+  (gnus-recent-update-message-id (alist-get 'message-id recent) to-group))
+
+(defun gnus-recent-forget-message-id (message-id &optional print-msg)
+  "Remove the Gnus article with MESSAGE-ID in `gnus-recent--articles-list'.
+When PRINT-MSG is non-nil, show a message about it."
+  (let ((l1 (length gnus-recent--articles-list))
+        (article (car gnus-recent--articles-list)))
+    ;; check for a match on the first article on list
+    (if (equal message-id (alist-get 'message-id article))
+        (pop gnus-recent--articles-list)
+      (setq article (gnus-recent-find-message-id message-id))
+      (cl-delete article gnus-recent--articles-list :test 'equal :count 1))
+    (when print-msg
+      (gnus-message 4 "Removed %d of 1 from gnus-recent articles" (- l1 (length gnus-recent--articles-list)))
+      (gnus-message 4 "Removed item: %s from gnus-recent articles" (car article)))))
 
 (defun gnus-recent-forget (recent &optional print-msg)
   "Remove RECENT Gnus article from `gnus-recent--articles-list'.
 When PRINT-MSG is non-nil, show a message about it."
-  (setq gnus-recent--articles-list
-        (cl-delete recent gnus-recent--articles-list :test 'equal :count 1))
-  (when print-msg
-    (message "Removed %s from gnus-recent articles" (car recent))))
+  (gnus-recent-forget-message-id (alist-get 'message-id recent) print-msg))
 
 (defun gnus-recent-forget-all (&rest _recent)
   "Clear the gnus-recent articles list."
   (interactive)
-  (setq gnus-recent--articles-list nil)
-  (message "Cleared all gnus-recent article entries"))
+  (when (yes-or-no-p "Action can not be undone. Are you sure? ")
+    (setq gnus-recent--articles-list nil)
+    (gnus-message 4 "Cleared all gnus-recent article entries")))
+
+(defun gnus-recent-bbdb-display-all (recent)
+  "Display sender and recipients in BBDB.
+Diplay sender and all recipients in BBDB. Ask to create a BBDB entry, if not in
+BBDB. RECENT is the gnus-recent data for the selected article."
+  (let ((recipients (alist-get 'recipients recent))
+        (search-list '(bbdb-search (bbdb-records))))
+    (setq recipients (append (bbdb-split "," (or (alist-get 'sender recent) ""))
+                             (bbdb-split "," (or (alist-get 'To recipients) ""))
+                             (bbdb-split "," (or (alist-get 'Cc recipients) ""))))
+    (dolist (r recipients)              ; add new entries to BBDB (ask)
+      (bbdb-update-records (list (list (gnus-recent-get-email-name r t)
+                                       (gnus-recent-get-email r t)))
+                           'query t))
+    ;; (bbdb-update-records                ; add new entries to BBDB (ask)
```

Comment:
Could you remove the dead code?

Reply:
Yes. 
But for the record it is not dead code, it is zombie code... lurking